ppa/archive/management/commands/generate_textcorpus.py

```diff
@@ -0,0 +1,200 @@
+
+    # Class attributes that rarely, if ever, need to change
+    DOC_ID_FIELD = "group_id_s"  # Solr field name for document identifier (not source_id, which is same across excerpts)
+    PAGE_ORDER_FIELD = "order"  # Solr field name for page ordering
+    OUTPUT_DOC_FIELDS = dict(
+        page_num_orig = 'label',
+        page_num_digi = 'order',
+        page_text = 'content',
+    )
+    PAGE_ID_FIELD = 'page_id'
+    WORK_ID_FIELD = 'work_id'
+    PAGE_NUM_FIELD = 'page_num_orig'
+    PAGE_SORT_FIELD = 'page_num_digi'
+
+    def __init__(self, path, doc_limit=-1):
+        """
+        A class encapsulating a Solr Client specification which yields
+       metadata and page data for PPA documents.
+
+        :param path: A string to a path for the corpus output.
+        :param doc_limit: Max no. of documents to process. The default of -1
+            means we process ALL documents found.
+        """
+        # root path for corpus
+        self.path = path
+
+        # limit docs queried
+        self.doc_limit = doc_limit
+
+        # subsequent paths
+        self.path_texts = os.path.join(self.path,'texts')
+        self.path_metadata = os.path.join(self.path,'metadata.json')
+        
+        # query to get initial results
+        results = SolrQuerySet().facet(self.DOC_ID_FIELD, limit=self.doc_limit)
+        # store page counts and doc ids
+        self.page_counts = results.get_facets().facet_fields[self.DOC_ID_FIELD]
+        self.doc_ids = self.page_counts.keys()
+        self.doc_count = len(self.doc_ids)
+    
+    @staticmethod
+    def _get_id(doc_id:str) -> str:
+        """Method to make a file-safe version of a document ID"""
+        return doc_id.replace('/','|')
+
+    def _get_meta_pages(self, doc_id:str) -> Tuple[dict,list]:
+        """Get metadata (dictionary) and pages (list of dictionaries) for a given document"""
+        
+        # get file safe work_id
+        work_id = self._get_id(doc_id)
+
+        # query
+        result = (
+            SolrQuerySet()
+            .search(**{self.DOC_ID_FIELD: doc_id})
+            .order_by(self.PAGE_ORDER_FIELD)
+        )
+
+        # populate the result cache with number of rows specified
+        docs = [
+            doc
+            for doc in result.get_results(rows=self.page_counts[doc_id])
+            if doc[self.DOC_ID_FIELD]==doc_id
+        ]
+
+        # find the metadata doc
+        metadata_docs = [d for d in docs if d["item_type"] == "work"]
+        assert len(metadata_docs)==1
+        meta = {self.WORK_ID_FIELD:work_id, **metadata_docs[0]}
+
+        # find the pages docs
+        page_docs = [d for d in docs if d["item_type"] == "page"]
+
+        # transform into new dictionary with keys in `self.PAGE_ID_FIELD` and `self.OUTPUT_DOC_FIELDS`
+        pages = [self._transform_doc(doc,meta) for doc in page_docs]
+
+        # make sure sorted by numeric page num (i.e. "digital")
+        pages.sort(key=lambda page: page[self.PAGE_SORT_FIELD])
```

Comment:
Is the Solr sorting unreliable?

Reply:
I found it was in one instance, yes, I can try to recreate.